src/main/java/org/truffleruby/extra/ffi/PointerNodes.java

```diff
@@ -47,13 +51,37 @@
 
     private static abstract class PointerPrimitiveArrayArgumentsNode extends PrimitiveArrayArgumentsNode {
 
-        private final BranchProfile nullPointerProfile = BranchProfile.create();
+        @Child private CheckNullPointerNode checkNullPointerNode;
 
         protected void checkNull(Pointer ptr) {
-            if (ptr.isNull()) {
+            if (checkNullPointerNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                checkNullPointerNode = insert(CheckNullPointerNode.create());
+            }
+
+            checkNullPointerNode.executeCheck(ptr);
+        }
+
+    }
+
+    @NodeChildren({ @NodeChild("pointer") })
+    public abstract static class CheckNullPointerNode extends RubyNode {
```

Comment:
This is not optimal for the footprint but probably not a big deal either.
One way to be better on footprint but still deduplicate code would be to define a static method taking a Pointer + BranchProfile.